web/src/app/components/bar-chart/bar-chart.component.ts

```diff
@@ -0,0 +1,341 @@
+import { formatDate } from '@angular/common'
+import * as d3 from 'd3';
+
+import { step189_2020 } from '../../../proto/step189_2020';
+
+/**
+ * BarChartData includes all data used by the bar chart.
+ */
+interface BarChartData {
+  pushID: string; // Push ID number
+  state: number;  // Tag of push end state
+  startTime: any; // Start time of push, in `YYYY-MM-DD HH:mm` format
+  duration: number; // Duration of push, in hour
+}
+
+type d3SVG = d3.Selection<SVGSVGElement, undefined, null, undefined>;
+type d3G = d3.Selection<SVGGElement, undefined, null, undefined>;
+type d3ScaleLinear = d3.ScaleLinear<number, number>;
+
+@Component({
+  selector: 'app-bar-chart',
+  templateUrl: './bar-chart.component.html',
+  styleUrls: ['./bar-chart.component.scss']
+})
+
+export class BarChartComponent implements OnChanges {
+  /**
+   * Private variables
+   * 
+   * The non-null assertion operator('!') suggests that there is no 
+   * issue of variables being `null` or `undefined`.
+   */
+  @ViewChild('barchart') private barChartContainer!: ElementRef; // Bar chart component
+  @Input() private pushInfos!: step189_2020.IPushInfo[] | null;
+  @Input() private currentPush!: step189_2020.IPushInfo | null;
+  private dataAll: BarChartData[] = [];
+  private dataComplete: BarChartData[] = [];
+  private totalDuration: number[] = [];
+  private completeDuration: number[] = [];
+  private svg!: d3SVG;
+  private focus!: d3G;
+  private brush!: d3G;
+  private brushSelector!: d3G;
+  private heightFocus!: number;
+  private heightBrush!: number;
+  private width!: number;
+  private xScaleFocus!: any;
+  private yScaleFocus!: d3ScaleLinear;
+  private xScaleBrush!: any;
+  private yScaleBrush!: d3ScaleLinear;
+  private xAxisFocus!: d3G;
+  private xAxisBrush!: d3G;
+  private yAxis!: d3G;
+
+  /**
+   * Constants
+   */
+  private static readonly NANO_TO_SECS: number = 10 ** 9;
+  private static readonly NANO_TO_MILLI: number = 10 ** 6;
+  private static readonly SECS_TO_HRS: number = 60 * 60;
+  private static readonly COLOR_LIGHT_GRAY: string = '#787878';
+  private static readonly COLOR_DARK_GRAY: string = '#373C38';
+  private static readonly COLOR_WHITE: string = '#eee';
+  private static readonly COLOR_WHITE_TRANS: string = '#ffffff00';
+  private static readonly DATE_FORMAT: string = 'yyyy-MM-dd HH:mm';
+  private static readonly DATE_LOCALE: string = 'en-US';
+  private static readonly COMPLETE_STATE_TAG: number = 5;
+  private static readonly BUILT_STATE_TAG: number = 14;
+  private static readonly BUILT_UNCHANGED_TAG: number = 17;
+  private static readonly SET_DEFAULT_VERSION_STATE_TAG: number = 18;
+  private static readonly ABANDONED_STATE_TAG: number = 19;
+  private static readonly STATE_TO_COLOR: { [index: number]: string } = {
+    5: '#34a853',
+    6: '#d50000',
+    10: '#2196f3',
+    12: '#d50000',
+    15: '#2196f3',
+    16: '#d50000'
+  };
+  
+  ngOnInit(): void { }
+
+  ngOnChanges(changes: SimpleChanges): void {
```

Comment:
Carol made a good point about this, but since we aren't feeding new data into the chart at any point on page reload, we can actually just use to `ngAfterViewInit` instead of `ngOnChanges` (psa just look at our code if you need help, it's really straightforward!)